### PR TITLE
Refactor create commands to split flags from options

### PR DESCRIPTION
**Pull request by lauchokyip**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Refactor commands to split flags from options

#### Which issue(s) this PR fixes:

Ref https://github.com/kubernetes/kubectl/issues/1046

#### Special notes for your reviewer:
1) I tried to add the `ResourceBuilder` flag to it but the `builder_flags.go` doesn't have the `Builder().Schema()` . Moreover, the filename flags in `builder_flags.go` clashed with [this Filename Flag](https://github.com/kubernetes/kubernetes/blob/aa450787bff54c96b4588e7ddbfdf9bda89a882a/staging/src/k8s.io/kubectl/pkg/cmd/create/create.go#L59) so I couldn't add it to it.
2) Couldn't remove [`PrintFlags`](https://github.com/kubernetes/kubernetes/blob/aa450787bff54c96b4588e7ddbfdf9bda89a882a/staging/src/k8s.io/kubectl/pkg/cmd/create/create.go#L51) and [`RecordFlags`](https://github.com/kubernetes/kubernetes/blob/aa450787bff54c96b4588e7ddbfdf9bda89a882a/staging/src/k8s.io/kubectl/pkg/cmd/create/create.go#L52) because it will be used [here](https://github.com/kubernetes/kubernetes/blob/aa450787bff54c96b4588e7ddbfdf9bda89a882a/staging/src/k8s.io/kubectl/pkg/cmd/create/create.go#L236) 
3) Removed the `TestExtraArgsFail` because `ToOptions` is combined with `Complete` and `Validate` now, so it will panic

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create.go [hunk 1]
```diff
@@ -46,6 +46,124 @@ import (
...
+	flags.PrintFlags.AddFlags(cmd)
+	flags.RecordFlags.AddFlags(cmd)
+
+	cmdutil.AddFilenameOptionFlags(cmd, &flags.FilenameOptions, "to use to create the resource")
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddApplyAnnotationFlags(cmd)
+	cmdutil.AddDryRunFlag(cmd)
+	cmdutil.AddFieldManagerFlagVar(cmd, &flags.fieldManager, "kubectl-create")
+
+	cmd.Flags().BoolVar(&flags.EditBeforeCreate, "edit", flags.EditBeforeCreate, "Edit the API resource before creating")
```
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create.go:91]: Let's get these move to util functions like above please.

**Comment by eddiezane** [comment]
Looking good! Not sure how I feel about the removal of complete + validate. We'll need to see more of these refactors before we commit on a pattern.

/hold

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create.go (hunk 1 above)
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create.go:91]: Hmm, I am not sure which one, is that `cmd.Flags().BoolVar()`? so I would have to create functions in `util` like `AddBoolVarFlags(cmd, "msg")`?

**Comment by dims** [comment]
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

**Comment by lauchokyip** [comment]
@ardaguclu and @soltysh , could you PTAL? Thanks!

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create.go [hunk 2]
```diff
@@ -65,7 +192,7 @@ type CreateOptions struct {
 	Raw              string
 
 	Recorder genericclioptions.Recorder
-	PrintObj func(obj kruntime.Object) error
+	Printer  printers.ResourcePrinter
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create.go:195]: What is the reason of this change in terms of options-flag split?

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_test.go [hunk 3]
```diff
@@ -35,13 +37,18 @@ func TestExtraArgsFail(t *testing.T) {
...
-	ioStreams, _, _, _ := genericclioptions.NewTestIOStreams()
-	options := NewCreateOptions(ioStreams)
-	if options.Complete(f, c, []string{"rc"}) == nil {
+	if validateCreateArgs(c, []string{"rc"}) == nil {
 		t.Errorf("unexpected non-error")
 	}
 }
 
+func validateCreateArgs(cmd *cobra.Command, args []string) error {
+	if len(args) != 0 {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_test.go:46]: As far as I can see, there are several validations happening in `ToOptions`. Would you mind adding more test cases to cover them?

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_test.go [hunk 4]
```diff
@@ -20,11 +20,13 @@ import (
 	"net/http"
 	"testing"
 
+	"github.com/spf13/cobra"
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_test.go:23]: nit: need a new line to separate cobra dependency from k8s one.

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create.go [hunk 5]
```diff
@@ -189,47 +290,6 @@ func (o *CreateOptions) Validate() error {
...
-	}
-	o.DryRunVerifier = resource.NewQueryParamVerifier(dynamicClient, f.OpenAPIGetter(), resource.QueryParamDryRun)
-	o.FieldValidationVerifier = resource.NewQueryParamVerifier(dynamicClient, f.OpenAPIGetter(), resource.QueryParamFieldValidation)
-
-	o.ValidationDirective, err = cmdutil.GetValidationDirective(cmd)
-	if err != nil {
-		return err
-	}
-
-	printer, err := o.PrintFlags.ToPrinter()
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create.go:221]: I think, these printer usage should stay as is in favor of this PR's simplicity.

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create.go [hunk 6]
```diff
@@ -46,6 +46,133 @@ import (
...
+	Recorder    genericclioptions.Recorder
+
+	fieldManager string
+
+	FilenameOptions  resource.FilenameOptions
+	Selector         string
+	EditBeforeCreate bool
+	Raw              string
+
+	genericclioptions.IOStreams
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create.go:63]: Why CreateFlags have iostreams?

Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create.go:63]: I was following the example for `wait` command https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go#L95

and besides that I think it makes sense to have `IOStreams` in `CreateFlags` since flags would rely on `stdout` `stdin` to process

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create.go (hunk 2 above)
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create.go:195]: I was following `wait` code, and I saw that it's using the generic `printers.ResourcePrinter` instead of `func(obj kruntime.Object) error`, so I looked into what the type definition of `printers.ResourcePrinter` 
```go
type ResourcePrinter interface {
	// PrintObj receives a runtime object, formats it and prints it to a writer.
	PrintObj(runtime.Object, io.Writer) error
}
```

so I guess `ResourcePrinter` is an almost similar but better options since it will take in an `io.Writer`, and you can write  to different output stream

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_test.go (hunk 3 above)
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_test.go:46]: I was looking at PR that got merged for split the flag but I couldn't find a good example , do you have any pointer where I can find a good example?

`TestExtraArgsFail` used to use `Complete` function which I removed so instead of removing the Test function, I just followed what `apply` command to test the `TestExtraArgsFail` https://github.com/kubernetes/kubernetes/blob/6bdda2da160043c5fcdfd25bbea9e1c1b804c9e5/staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go#L95

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_test.go (hunk 4 above)
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_test.go:23]: Fixed, thanks!

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR
